coderd/workspaces.go

```diff
@@ -178,8 +179,25 @@ func (api *API) workspaces(rw http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	var filteredWorkspaces []codersdk.Workspace
+	// apply post filters, if they exist
+	if postFilter.DeletingBy == nil {
+		filteredWorkspaces = append(filteredWorkspaces, wss...)
+	} else {
+		for _, v := range wss {
+			if v.DeletingAt == nil {
+				continue
+			}
+			// get the beginning of the day on which deletion is scheduled
+			truncatedDeletionAt := v.DeletingAt.Truncate(24 * time.Hour)
```

Comment:
@johnstcn I've spent a while on this and I'm a bit stuck.
I could see two reasonable paths forward:
1) The FE handles timezoning and just sends down a UTC date. The BE logic remains the same. Maybe this makes testing a little brittle.
2) The FE sends down a date-time string with an offset that can be easily parsed in Go, as you suggested in Slack. For example, the FE could send down `"Mon Jan 02 15:04:05 MST 2006"`.  In `search.go`, we could parse the param: 
````
postFilter.DeletingBy = ptr.Ref(parser.Time(values, time.Time{}, "deleting_by", time.UnixDate))
````
Then, we could calculate the above as follows:
````
truncatedDeletionAt := time.Date(v.DeletingAt.Year(), v.DeletingAt.Month(), v.DeletingAt.Day(), 0, 0, 0, 0, postFilter.DeletingBy.Location())
````

Here's the problem with both of these approaches: I can't send down times delimited with a colon (`:`) because our filtering API uses colons for parsing params. I really am not trying to rewrite the parser at the moment.

Alternatively, I've tried sending down a completely distinct param:
````
@Param tz query string false "Specify a IANA time zone name to be used with deleting_by param"
````
With this param, I could use `time.LoadLocation(postFilter.TZ)` to get location; however, currently all our query params are lower-cased, which screws up the `tz` naming. Of course I could write some code to bypass this measure for this specific param, but that just seems gross to me.

Is there a simpler approach I'm missing?

Reply:
Okay, for posterity's sake, adding the discussion via email:

> I've spent a while on this and I'm a bit stuck.
> I could see two reasonable paths forward:
> 
> The FE handles timezoning and just sends down a UTC date. The BE logic remains the same. Maybe this makes testing a little brittle.
> The FE sends down a date-time string with an offset that can be easily parsed in Go, as you suggested in Slack. For example, the FE could send down `"Mon Jan 02 15:04:05 MST 2006"`. In `search.go`, we could parse the param:
> ````
> postFilter.DeletingBy = ptr.Ref(parser.Time(values, time.Time{}, "deleting_by", time.UnixDate))
> ````
> Then, we could calculate the above as follows:
> ````
> truncatedDeletionAt := time.Date(v.DeletingAt.Year(), v.DeletingAt.Month(), v.DeletingAt.Day(), 0, 0, 0, 0, postFilter.DeletingBy.Location())
> ````
> Here's the problem with both of these approaches: I can't send down times delimited with a colon (`:`) because our filtering API uses colons for parsing params. I really am not trying to rewrite the parser at the moment.
> 
> Alternatively, I've tried sending down a completely distinct param:
> 
> `@Param tz query string false "Specify a IANA time zone name to be used with deleting_by param"`
> With this param, I could use `time.LoadLocation(postFilter.TZ)` to get location; however, currently all our query params are lower-cased, which screws up the tz naming. Of course I could write some code to bypass this measure for this specific param, but that just seems gross to me.
> 
> Is there a simpler approach I'm missing?

> Both of the issues you describe above are annoying.
> Spike and I had a quick chat about this and we came up with a few ideas:
> 1) Use a custom time layout. Go will happily accept any time layout you use, so you could use e.g. "2006-01-02-0700" to parse date/time strings like "2023-06-12+0530". The FE could still handle timezoning but it would send the date in that custom format that doesn't interfere with the query parser. This is also relatively intuitive for a user to manually edit in the query string. You might need a custom type to handle the JSON marshal/unmarshaling properly though, which isn't terrible but is slightly annoying to deal with.
> 2) Use a UNIX timestamp to represent the date instead. Parsing this in Go is trivial with `time.Unix()` and comparing times with `t.Equal(u)` method will compare the instants and ignore the locations. This is also less friendly for users to type in unless there's some kind of UI support e.g. a date picker.
> 3) Send the timezone offset to the backend as a separate query parameter (that is, it wouldn't show up in the query box at all and would be handled outside of the query parser). This would make it work 'magically' but with the caveat that if you copy + paste the URL to someone in a different timezone they might not see what you see. This could be as simple as adding a new optional string parameter to the workspaces endpoint, falling back to UTC offset if none is specified. 
> 4) Ship what you have right now with the timezone caveat, and open a follow-up issue to fix the query parser to be able to handle RFC3339-formatted dates. 
> Option 4 (fixing the query parser) is probably the highest benefit as it would allow us to properly handle dates/times both now and in the future for other fields.

Conclusion:
We are going with option 4) as the audit log filter suffers from the same issue. Ticket for that future work [here](https://github.com/coder/coder/issues/7973).